### ISSUE TITLE
Variable are not properly marked as used when modules or globalReturn are enabled

### ISSUE THREAD

**Issue by MoOx** [yannickcr/eslint-plugin-react#18]
`src/widget.js`

```jsx
"use strict";

var React = require("react")
var HelloWorld = require("op/HelloWorld")

React.render(
  <div className="op-Widget">
    <HelloWorld />
  </div>,
  document.getElementById("op-Widget")
)
```

`.eslintrc`

```yaml
ecmaFeatures:
  modules: true
  # spread: true
  jsx: true

env:
  es6: true
  browser: true
  node: true

globals:
  __DEV__: true
  __PROD__: true

plugins:
  - react

# 0: off, 1: warning, 2: error
rules:
  # semicolons are useless
  semi: 0

  # @todo use the line below
  quotes: [2, "double"]

  # 2 spaces indentation,
  indent: [2, 2]

  # trailing coma are cool for diff
  comma-dangle: [2, "always"]

  # enforce comma at eol (never before)
  comma-style: [2, "last"]

  valid-jsdoc: 1

  # eslint-plugin-react rules
  react/no-multi-comp: 2
  react/prop-types: 2
  react/display-name: 2
  react/wrap-multilines: 2
  react/self-closing-comp: 2
  react/no-did-mount-set-state: 2
  react/no-did-update-set-state: 2
  react/jsx-uses-react: 2
  react/react-in-jsx-scope: 2

```

Gives

```console
$ eslint src/widget.js

src/widget.js
   8:4  error  HelloWorld is defined but never used     no-unused-vars
  11:2  error  'React' must be in scope when using JSX  react/react-in-jsx-scope
  12:4  error  'React' must be in scope when using JSX  react/react-in-jsx-scope
  12:5  error  'HelloWorld' is not defined              no-undef

✖ 4 problems (4 errors, 0 warnings)
```

What am I missing ?

**Comment by MoOx** [comment]
FYI I am using eslint 0.16.2

**Comment by MoOx** [comment]
Maybe related to https://github.com/eslint/eslint/issues/2004
Not sure.

**Comment by MoOx** [comment]
It seems related to https://github.com/eslint/eslint/issues/1911

**Comment by MoOx** [comment]
So the PR that would allow you to fix that is probably https://github.com/eslint/eslint/pull/1951

**Comment by yannickcr** [comment]
Yes, it seems there is some problems with the JSX variables handling in ESLint 0.16

Also with eslint/eslint#1951 ESLint will stop to handle JSX variables and will rely on a rule in `eslint-plugin-react` to handle them.

I'll try to figure out the issue in ESLint 0.16 and see how we can fix it.

Fox now you can stay on ESLint 0.15, there is no issues with it.

**Comment by MoOx** [comment]
Good to hear.

> Fox now you can stay on ESLint 0.15, there is no issues with it.

Nope since it don't handle module :)
I just deactivated the rules & will enable them back after that.

**Comment by yannickcr** [comment]
As said earlier, ESLint stopped handling JSX variables in v0.17.0 ([see details](http://eslint.org/blog/2015/03/eslint-0.17.0-released/#changes-to-jsx/react-handling)).

I added a new rule in v1.5.0 to handle JSX variables ([jsx-uses-vars](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-uses-vars.md)).

So, to fix this issue:

- update to ESLint v0.17.0
- update to eslint-plugin-react v1.5.0
- add `"react/jsx-uses-vars":2` to your `.eslintrc` config file

And it should be fine :)

**Comment by zonak** [comment]
I am still seeing the same issue :(

I have the following installed:

- eslint@0.17.0
- eslint-plugin-react@1.5.0

Have the following `.eslintrc`:
```json
{
  "env": {
    "browser": true,
    "node": true,
    "es6": true
  },
  "ecmaFeatures": {
    "modules": true,
    "jsx": true
  },
  "plugins": [
    "react"
  ],
  "rules": {
    "strict": 0,
    "no-throw-literal": 1,
    "quotes": [1, "single"],
    "react/no-multi-comp": 1,
    "react/prop-types": 1,
    "react/display-name": 1,
    "react/wrap-multilines": 1,
    "react/self-closing-comp": 1,
    "react/no-did-mount-set-state": 1,
    "react/no-did-update-set-state": 1,
    "react/jsx-uses-react": 1,
    "react/jsx-uses-vars": 2,
    "react/react-in-jsx-scope": 1
  }
}
```

Using the following sample code:

```js
import React from 'react';

let HelloMessage = React.createClass({
  render: function() {
    return <div>Hello {this.props.name}!!!</div>;
  }
});

React.render(<HelloMessage name="zonak" />, document.getElementById('react-container'));
```
getting the following errors:

```
line 3, col 4, Error - HelloMessage is defined but never used (no-unused-vars)
line 3, col 37, Warning - 'name' is missing in props validation (react/prop-types)
line 3, col 37, Warning - Component definition is missing display name (react/display-name)
line 9, col 14, Error - 'HelloMessage' is not defined. (no-undef)
```

**Comment by yannickcr** [comment]
Ok, seems there is a problem marking a variable as used when the ecmaFeatures `modules` or `globalReturn` are enabled. Investigating.

**Comment by zigomir** [comment]
I think I have same issue. I'm also using `modules` in `ecmaFeatures`.

**Comment by yannickcr** [comment]
Will be fixed when eslint/eslint#2121 gets merged.

**Comment by yannickcr** [comment]
Fixed in [ESLint 0.18.0](https://github.com/eslint/eslint/releases/tag/v0.18.0)
